src/Form/MetadataDisplayForm.php

```diff
@@ -11,69 +19,205 @@
+      '#type' => 'button',
+      '#op' => 'preview',
+      '#value' => $this->t('Show preview'),
+      '#ajax' => [
+        'callback' => [$this, 'ajaxPreview'],
+      ],
+      '#states' => [
+        'visible' => ['input[name="ado_context_preview"' => ['filled' => true]],
+      ],
+    ];
+
+    // Enable autosaving in code mirror.
+    $form['#attached']['library'][] = 'format_strawberryfield/code_mirror_autosave';
+
     return $form;
   }
 
+  /**
+   * {@inheritdoc}
+   */
   public function validateForm(array &$form, FormStateInterface $form_state) {
     try {
       $build = [
         '#type' => 'inline_template',
         '#template' => $form_state->getValue('twig')[0]['value'],
         '#context' => [],
       ];
-      \Drupal::service('renderer')->renderPlain($build);
+      $this->renderer->renderPlain($build);
     }
     catch (\Exception $exception) {
-      $form_state->setErrorByName('twig',$exception->getMessage());
+      $form_state->setErrorByName('twig', $exception->getMessage());
     }
-
-    return parent::validateForm(
-      $form,
-      $form_state
-    ); // TODO: Change the autogenerated stub
+    //@TODO Change the autogenerated stub.
+    return parent::validateForm($form, $form_state);
   }
 
-
   /**
    * {@inheritdoc}
    */
   public function save(array $form, FormStateInterface $form_state) {
     $status = parent::save($form, $form_state);
 
+    /** @var \Drupal\format_strawberryfield\MetadataDisplayInterface $entity */
     $entity = $this->entity;
     if ($status == SAVED_UPDATED) {
       $this->messenger()->addMessage($this->t('The Metadata Display %entity has been updated.', ['%entity' => $entity->toLink()->toString()]));
-    } else {
+    }
+    else {
       $this->messenger()->addMessage($this->t('The Metadata Display %entity has been added.', ['%entity' => $entity->toLink()->toString()]));
     }
-    \Drupal::service('plugin.manager.field.formatter')->clearCachedDefinitions();
-
-    /** @var \Drupal\Core\Template\TwigEnvironment $environment */
-    $environment = \Drupal::service('twig');
-    $environment->invalidate();
+    $this->formatterPluginManager->clearCachedDefinitions();
+    $this->twig->invalidate();
 
-    $form_state->setRedirectUrl($this->entity->toUrl('collection'));
     return $status;
   }
+
+  /**
+   * AJAX callback.
+   */
+  public static function ajaxPreview($form, FormStateInterface $form_state) {
+    $response = new AjaxResponse();
+
+    /** @var \Drupal\format_strawberryfield\MetadataDisplayInterface $entity */
+    $entity = $form_state->getFormObject()->getEntity();
+
+    // Attach the library necessary for using the OpenOffCanvasDialogCommand and
+    // set the attachments for this Ajax response.
+    $form['#attached']['library'][] = 'core/drupal.dialog.off_canvas';
+    $form['#attached']['library'][] = 'codemirror_editor/editor';
+    $response->setAttachments($form['#attached']);
+
+    $input = $form_state->getUserInput();
```

Comment:
Question, is the `ado_context_preview` not in the $form-state as value? Wonder if getting data from the User Input, not validated may bring issues? (testing that now)

Reply:
That's because on ajax requests, multisteps etc, the values are not guaranteed to be there. As the user is already a privileged one I don't think this is going to bring much issues. It is a good point that we're not validating the twig itself so we might want to do that? not sure if it's worth it.

Reply:
We are validating the twig at the entity level. Here: https://github.com/esmero/format_strawberryfield/blob/main/src/Entity/MetadataDisplayEntity.php#L278

and also catching render exceptions (which is an implicit Twig template validation) here https://github.com/esmero/format_strawberryfield/pull/129/files/b4ce99ee2bd874025b193b4581dc611e18bea926#diff-376335f1b2fb78915a415da3636afd7ce6a2927d608c5bda585db4cbce30d91fR124

I can concord with not a strict validation if $form_state is unreliable because of the Code Mirror and it seems like ::extractEntityIdFromAutocompleteInput by doing regular expression may be enough. I will roll with getUserInput() as you have it now. Thanks